fe_utils/finite_elements.py

```diff
@@ -37,7 +42,20 @@ def vandermonde_matrix(cell, degree, points, grad=False):
     <ex-vandermonde>`.
     """
 
-    raise NotImplementedError
+    def column(i, idx, j):  # computes single column of vandermonde matrix
+        exp_1 = max([(i - j) - grad * (1 - idx), 0])  # (i-j) for grad=False
+        exp_2 = max([j - grad * idx, 0])  # j for grad=False
+        return np.multiply(points[:, None, 0] ** exp_1, points[:, None, -1] ** exp_2)
+
+    matrix = []
+    if grad:  # i loops over the degrees, j loops over the columns containing the polynomials of degree at most i
+        for i in range(degree + 1):
+            matrix += [np.hstack([((i - j) * (1 - k) + j * k) * column(i, k, j) for k in range(cell.dim)])
+                       for j in range((cell.dim - 1) * i + 1)]  # k loops over the dimensions of a point
+        return np.swapaxes(matrix, 0, 1)
+    for i in range(degree + 1):
+        matrix += [column(i, 0, j) for j in range((cell.dim - 1) * i + 1)]
+    return np.hstack(matrix)
```

Comment:
This is quite a nice way of doing this, subject again to the unnecessary verbosity of writing loops instead of comprehensions.